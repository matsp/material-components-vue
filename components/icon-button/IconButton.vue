<template>
  <a
    v-if="href !== ''"
    :href="href"
    :data-toggle-on-content="toggleOnContent"
    :data-toggle-on-label="toggleOnLabel"
    :data-toggle-on-class="toggleOnClass"
    :data-toggle-off-content="toggleOffContent"
    :data-toggle-off-label="toggleOffLabel"
    :data-toggle-off-class="toggleOffClass"
    v-bind="$attrs"
    class="mdc-icon-button"
    :class="classes"
    v-on="$listeners"
    @MDCIconButtonToggle:change="$emit('change', $event.detail.isOn)"
  >
    <slot v-if="isIconButton" />
    <slot
      v-if="isToggleButtonViaSlots && !value"
      name="toggleOn"
    />
    <slot
      v-if="isToggleButtonViaSlots && value"
      name="toggleOff"
    />
  </a>
  <button
    v-else
    :data-toggle-on-content="toggleOnContent"
    :data-toggle-on-label="toggleOnLabel"
    :data-toggle-on-class="toggleOnClass"
    :data-toggle-off-content="toggleOffContent"
    :data-toggle-off-label="toggleOffLabel"
    :data-toggle-off-class="toggleOffClass"
    v-bind="$attrs"
    class="mdc-icon-button"
    :class="classes"
    v-on="$listeners"
    @MDCIconButtonToggle:change="$emit('change', $event.detail.isOn)"
  >
    <template v-if="this.icon && this.icon !== ''">
      {{ icon }}
    </template>
    <slot v-if="isIconButton" />
    <slot
      v-if="isToggleButtonViaSlots && !value"
      name="toggleOn"
    />
    <slot
      v-if="isToggleButtonViaSlots && value"
      name="toggleOff"
    />
  </button>
</template>

<script>
import { MDCRipple } from '@material/ripple'
import { MDCIconButtonToggle } from '@material/icon-button'

import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  model: {
    prop: 'value',
    event: 'change'
  },
  props: {
    toggleOnContent: {
      type: String,
      default: ''
    },
    toggleOnLabel: {
      type: String,
      default: ''
    },
    toggleOnClass: {
      type: String,
      default: ''
    },
    toggleOffContent: {
      type: String,
      default: ''
    },
    toggleOffLabel: {
      type: String,
      default: ''
    },
    toggleOffClass: {
      type: String,
      default: ''
    },
    value: {
      type: Boolean,
      default: false
    },
    href: {
      type: String,
      default: ''
    },
    icon: {
      type: String,
      default: ''
    }
  },
  data () {
    return {
      mdcIconButtonToggle: undefined,
      mdcRipple: undefined,
      slotObserver: undefined
    }
  },
  computed: {
    isToggleButtonViaSlots () {
      return this.$slots.toggleOn && this.$slots.toggleOff && !this.isIconButton
    },
    isIconButton () {
      return this.$slots.default
    },
    isToggleButton () {
      return this.toggleOnContent !== '' && this.toggleOffContent !== ''
    },
    classes () {
      return {
        'material-icons': this.icon && this.icon !== ''
      }
    }
  },
  watch: {
    value (value) {
      if (typeof this.mdcIconButtonToggle !== 'undefined') {
        this.mdcIconButtonToggle.on = value
      }
    }
  },
  mounted () {
    this.update()
    this.slotObserver = new MutationObserver(() => this.update())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
  },
  beforeDestroy () {
    if (typeof this.mdcIconButtonToggle !== 'undefined') {
      this.mdcIconButtonToggle.destroy()
    }
    if (typeof this.mdcRipple !== 'undefined') {
      this.mdcRipple.destroy()
    }
    this.slotObserver.disconnect()
  },
  methods: {
    update () {
      if ((this.isToggleButton || this.isToggleButtonViaSlots) && typeof this.mdcIconButtonToggle === 'undefined') {
        this.mdcIconButtonToggle = MDCIconButtonToggle.attachTo(this.$el)
        this.mdcIconButtonToggle.on = this.value
      }
      if (this.isIconButton && typeof this.mdcRipple === 'undefined') {
        this.mdcRipple = MDCRipple.attachTo(this.$el)
        this.mdcRipple.unbounded = true
      }
    }
  }
}
</script>
