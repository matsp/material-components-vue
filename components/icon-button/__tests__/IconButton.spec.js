import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import IconButton from '../IconButton.vue'

describe('IconButton', () => {
  it('should mount', () => {
    const wrapper = mount(IconButton)
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.vm.$data.slotObserver).toBeDefined()
  })

  it('should render with no prop', () => {
    const wrapper = mount(IconButton)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-icon-button')
  })
})
