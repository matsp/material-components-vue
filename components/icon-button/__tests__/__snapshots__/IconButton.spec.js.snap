// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton should render with no prop 1`] = `
<button data-toggle-on-content="" data-toggle-on-label="" data-toggle-on-class="" data-toggle-off-content="" data-toggle-off-label="" data-toggle-off-class="" class="mdc-icon-button">
  <!---->
  <!---->
  <!---->
  <!----></button>
`;
